### PR TITLE
Update readme

### PR DESCRIPTION
**Pull request by cbourjau**
This PR makes the readme more concise and focuses on the information most
relevant to an impatient new user. I also added a credits section.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,82 +1,65 @@
...
-  - Checks are performed as eagerly as possible!
-- Supports the entirety of modern opsets, including features like subgraphs (control flow) and types other than tensors (like sequences and optionals).
-  - Standard operators all have typed Python signatures and docstrings!
-- Is designed for predictability. No mutable types are passed around, so it's difficult to invalidate the graph accidentally.
-  - If it's legal Spox, it (should be) legal ONNX!
-
-The main goal of Spox is to provide a robust and Pythonic framework for developing libraries building ONNX graphs, such as converters or other custom applications.
+Converting a trained model into ONNX entails replicating its inference logic with ONNX operators.
+In the past that constituted a major undertaking.
+Based on those experiences we designed Spox from the ground up to make the process of writing converters (and ONNX models in general) as easy as possible.
```
Comment (jbachurski) [inline README.md:12]: We're missing the actual _why_ right now, only stating that we make it easy.

```suggestion
A common application of ONNX is converting models from various frameworks. This requires replicating its runtime behaviour with ONNX operators.
In the past this has been a major challenge.
Based on our experience, we designed Spox from the ground up to make the process of writing converters (and ONNX models in general) as easy as possible. 

Spox's features include:

- Eager validation and type inference (as the graph is constructed)
- Errors with Python tracebacks to offending operators
- First-class support for subgraphs (control flow)
- A lean and predictable API
```

File: README.md [hunk 2]
```diff
@@ -1,82 +1,65 @@
...
 
 ## Installation
 
-Spox is published on conda-forge and can be installed as expected:
+With pip:
```
Comment (jbachurski) [inline README.md:16]:
```suggestion
Spox releases are available on PyPI:
```

File: README.md [hunk 3]
```diff
@@ -1,82 +1,65 @@
...
-You may print out the `Var` or check its `Var.type` to learn a bit more about it.
-
-To perform operations on a `Var`, use an _opset_ (operator set) module like `spox.opset.ai.onnx.v17` or `spox.opset.ai.onnx.ml.v3` - which correspond to `ai.onnx@17` and `ai.onnx.ml@3`, the standard opsets. These are pre-generated for you and you may import them as required.
-
-For instance, using the default opset you could write a function that given two variables returns their [geometric mean](https://en.wikipedia.org/wiki/Geometric_mean):
-
-```python
-from spox import Var
-from spox.opset.ai.onnx import v17 as op
+Spox is also available on conda-forge and can be installed as expected:
```
Comment (jbachurski) [inline README.md:22]:
```suggestion
There is also a package available on conda-forge:
```
Also, technically this isn't quite there yet.

File: README.md [hunk 4]
```diff
@@ -1,82 +1,65 @@
...
-    )
+```bash
+conda install spox
 ```
 
-Since ONNX is tensor-oriented, this code assumes that `x` and `y` are floating point tensors with matching (broadcastable) shapes, and the geometric mean is computed elementwise. You may learn more about what types and shapes are acceptable for given operators by checking their docstring.
+## Quick start
 
-Performing operations on Variables creates further Variables, all of which keep track of what computation has been performed so far. If an operation is determined to be illegal (for example due to mismatching types/shapes), an exception will be immediately raised by Spox.
+Spox users interact most commonly with `Var` objects - **variables** which will have values assigned to them at runtime.
```
Comment (jbachurski) [inline README.md:30]:
```suggestion
In Spox, you primarily interact with `Var` objects - **variables** - which are placeholders for runtime values.
```

File: README.md [hunk 5]
```diff
@@ -1,82 +1,65 @@
...
+The initial `Var` objects which represent the *argument*s of a model (the model inputs in ONNX nomenclature) are created with an explicit type and shape using the `spox.argument` function.
```
Comment (jbachurski) [inline README.md:31]:
```suggestion
The initial `Var` objects, which represent the *arguments* of a model (the model inputs in ONNX nomenclature), are created with an explicit type using the `argument(Type) -> Var` function. The possible types include `Tensor`, `Sequence`, and `Optional`.
```

File: README.md [hunk 6]
```diff
@@ -1,82 +1,65 @@
...
+All further `Var` objects are created by calling functions which take existing `Var` objects as inputs and produce new `Var` objects as outputs.
+Spox provides such functions for all operators in the standard grouped by domain and version inside the `spox.opset` module.
```
Comment (jbachurski) [inline README.md:33]:
```suggestion
Spox provides such functions - _operator constructors_ - for all operators in the standard grouped by domain and version inside the `spox.opset` module.
```

File: README.md [hunk 7]
```diff
@@ -1,82 +1,65 @@
...
 
-### Building
+The final `onnx.ModelProto` object is build by specifying the explicit input and output `Var`s of the model to in the `spox.build` function.
```
Comment (jbachurski) [inline README.md:35]: A bit hard to parse right now.
```suggestion
The final `onnx.ModelProto` object is built by passing input and output `Var`s for the model to the `spox.build` function.
```

File: README.md [hunk 8]
```diff
@@ -1,82 +1,65 @@
...
 
-To introduce _argument_ variables (which are placeholders for runtime values - graph/model inputs in ONNX nomenclature), you may use the `argument(typ: Type) -> Var` function with the required type, like `Tensor`.
-
-When you're done you may build an `onnx.ModelProto` - the model protobuf, which is the ONNX program representation. You may do this with the public `build(inputs, outputs) -> onnx.ModelProto` function:
+Below is an example for defining an ONNX graph which computes the [geometric mean](https://en.wikipedia.org/wiki/Geometric_mean) of two inputs.
+Please consult the [documentation](https://spox.readthedocs.io/en/latest) for more details and further tutorials on how to use Spox.
```
Comment (jbachurski) [inline README.md:38]:
```suggestion
Make sure to consult the Spox [documentation](https://spox.readthedocs.io/en/latest) to find more details and tutorials.
```

File: README.md [hunk 9]
```diff
@@ -1,82 +1,65 @@
...
+# and has the runtime-defined length "N".
+a = argument(Tensor(float, ('N',)))
+b = argument(Tensor(float, ('N',)))
 
+# Perform operations on `Var`s
 c = geometric_mean(a, b)
 
-model = build({'a': a, 'b': b}, {'c': c})
+# Build an `onnx.ModelProto` for the given inputs and outputs.
+model: onnx.ModelProto = build(inputs={'a': a, 'b': b}, outputs={'c': c})
```
Comment (jbachurski) [inline README.md:60]: I don't see why writing out the argument names is necessary? It already goes left-to-right, so inputs-then-outputs seems intuitive enough.
```suggestion
model: onnx.ModelProto = build({'a': a, 'b': b}, {'c': c})
```

File: README.md [hunk 10]
```diff
@@ -1,82 +1,65 @@
...
+conda install spox
 ```
 
-Since ONNX is tensor-oriented, this code assumes that `x` and `y` are floating point tensors with matching (broadcastable) shapes, and the geometric mean is computed elementwise. You may learn more about what types and shapes are acceptable for given operators by checking their docstring.
+## Quick start
 
-Performing operations on Variables creates further Variables, all of which keep track of what computation has been performed so far. If an operation is determined to be illegal (for example due to mismatching types/shapes), an exception will be immediately raised by Spox.
+Spox users interact most commonly with `Var` objects - **variables** which will have values assigned to them at runtime.
+The initial `Var` objects which represent the *argument*s of a model (the model inputs in ONNX nomenclature) are created with an explicit type and shape using the `spox.argument` function.
+All further `Var` objects are created by calling functions which take existing `Var` objects as inputs and produce new `Var` objects as outputs.
```
Comment (jbachurski) [inline README.md:32]:
```suggestion
All further `Var` objects are created by calling functions which take existing `Var` objects as inputs and produce new `Var` objects as outputs. Spox determines the `Var.type` for these eagerly to allow validation.
```

File: README.md [hunk 11]
```diff
@@ -1,82 +1,65 @@
...
-
-When you're done you may build an `onnx.ModelProto` - the model protobuf, which is the ONNX program representation. You may do this with the public `build(inputs, outputs) -> onnx.ModelProto` function:
+Below is an example for defining an ONNX graph which computes the [geometric mean](https://en.wikipedia.org/wiki/Geometric_mean) of two inputs.
+Please consult the [documentation](https://spox.readthedocs.io/en/latest) for more details and further tutorials on how to use Spox.
 
 ```python
+import onnx
+
 from spox import argument, build, Tensor, Var
+# import operators from the ai.onnx domain at version 17
```
Comment (jbachurski) [inline README.md:44]:
```suggestion
# Import operators from the ai.onnx domain at version 17
```

File: README.md [hunk 12]
```diff
@@ -1,82 +1,65 @@
...
 
-def geometric_mean(x: Var, y: Var) -> Var: ...
+def geometric_mean(x: Var, y: Var) -> Var:
+    # use the standard Sqrt and Mul
+    return op.sqrt(op.mul(x, y))
 
-a = argument(Tensor(float, (1, 'N')))
-b = argument(Tensor(float, ('M', 1)))
+# Create typed model inputs. Each tensor is of rank 1
+# and has the runtime-defined length "N".
```
Comment (jbachurski) [inline README.md:52]: I wouldn't say it's _defined_ by a runtime.
```suggestion
# and has the runtime-determined length 'N'.
```

**Review by jbachurski — CHANGES_REQUESTED** [review]
I do have some comments to fix.

File: README.md (hunk 3 above)
Comment (cbourjau) [reply README.md:22]: Yes, coming up.

File: README.md (hunk 6 above)
Comment (cbourjau) [reply README.md:33]: My thought here: It does not matter to the user what is happening inside the functions. Why introduce the term "operator constructor" at all?

File: README.md (hunk 9 above)
Comment (cbourjau) [reply README.md:60]: I think being a bit more explicit in the example does not hurt.

File: README.md (hunk 6 above)
Comment (jbachurski) [reply README.md:33]: Most documentation articles use that term, including the listings. I think it's just convenient to call these things somehow, as they have an important role.

File: README.md (hunk 9 above)
Comment (jbachurski) [reply README.md:60]: Fair enough

File: README.md (hunk 6 above)
Comment (cbourjau) [reply README.md:33]: But from the user's point of view it makes no difference if you are dealing with your own functions or with a "constructor function" from spox. In any case, I don't think we need to introduce the name here in the readme.

File: README.md [hunk 13]
```diff
@@ -1,82 +1,65 @@
...
-
-Building the model may be done for you if you're writing a component of a library using Spox.
-
-### Running
-
-Afterwards, you may use an ONNX runtime/backend to execute the model with some inputs, for example the [ONNX Runtime](https://github.com/microsoft/onnxruntime).
+## Credits
 
-You can learn more about Spox in the [documentation](https://docs.dev.quantco.cloud/qc-github-artifacts/Quantco/spox/latest/index.html).
+Original authors: [@jbachurski](https://github.com/jbachurski) with supervision from [@cbourjau](https://github.com/cbourjau).
```
[the remark this answers is not in the document]
Comment (cbourjau) [reply README.md]: I am closing this as discussed on slack.

File: README.md [hunk 14]
```diff
@@ -1,82 +1,72 @@
...
 
-- Implements the ONNX standard while also allowing Pythonic code.
-  - Spox follows conventions set out by numpy and other Python libraries.
-- Enforces the strong type system of ONNX, by raising errors with Python tracebacks to the offending operator.
-  - Checks are performed as eagerly as possible!
-- Supports the entirety of modern opsets, including features like subgraphs (control flow) and types other than tensors (like sequences and optionals).
-  - Standard operators all have typed Python signatures and docstrings!
-- Is designed for predictability. No mutable types are passed around, so it's difficult to invalidate the graph accidentally.
-  - If it's legal Spox, it (should be) legal ONNX!
+A common application of ONNX is converting models from various frameworks. This requires replicating its runtime behaviour with ONNX operators.
```
Comment (jbachurski) [inline README.md:10]:
```suggestion
A common application of ONNX is converting models from various frameworks. This requires replicating their runtime behaviour with ONNX operators.
```

**Review by JakubBachurskiQC — APPROVED** [review]
